test/integration/middleware-dev-errors/test/index.test.js

```diff
@@ -241,4 +242,59 @@ describe('Middleware development errors', () => {
       expect(await browser.elementByCss('#page-title')).toBeTruthy()
     })
   })
+
+  describe('when there is a compilation error from boot', () => {
+    beforeEach(() => {
+      context.middleware.write(`export default function () }`)
+    })
+
+    it('logs the error correctly', async () => {
+      await fetchViaHTTP(context.appPort, '/')
+      expect(context.logs.output).toContain(`Expected '{', got '}'`)
+      expect(context.logs.output.split(`Expected '{', got '}'`).length).toEqual(
+        2
+      )
+    })
+
+    it('renders the error correctly and recovers', async () => {
+      const browser = await webdriver(context.appPort, '/')
+      expect(await hasRedbox(browser, true)).toBe(true)
+      expect(
+        await browser
+          .elementByCss('#nextjs__container_build_error_label')
+          .text()
+      ).toEqual('Failed to compile')
+      context.middleware.write(`export default function () {}`)
+      await hasRedbox(browser, false)
+      expect(await browser.elementByCss('#page-title')).toBeTruthy()
+    })
+  })
+
+  describe('when there is a compilation error after boot', () => {
+    beforeEach(() => {
+      context.middleware.write(`export default function () {}`)
+    })
+
+    it('logs the error correctly', async () => {
+      await fetchViaHTTP(context.appPort, '/')
+      context.middleware.write(`export default function () }`)
+      await check(() => {
+        expect(context.logs.output).toContain(`Expected '{', got '}'`)
```

Comment:
Does `expect` inside of the `check` properly retry or does `jest` track the failure here and still fail the test?

Reply:
I believe `expect` would throw if the expectation is not satisfied so `check` would retry until both success.